TijanaSonjic/assignment3/wordSearch.cpp

```diff
@@ -0,0 +1,160 @@
+#include "test_libraryr.h"
+
+using namespace std;
+
+
+
+set<string> dictionary = { "apple", "spring", "sun", "sea", "dancer" };
+
+typedef struct gridCell {
+	char letter;
+	bool visited;
+} GridCell;
+
+typedef struct delta {
+	int dx;
+	int dy;
+} Delta;
+
+GridCell grid[][3] = {
+						{ {'s', false}, {'p', false}, {'g', false} },
+						{ {'t', false}, {'r', false}, {'n', false} },
+						{ {'s', false}, {'u', false}, {'i', false} }
+					};
+
+// Directions for adjacent cells
+Delta difference[8] = { {-1,-1},{-1,0},{-1,1},{0,-1},{0,1},{1,-1},{1,0},{1,1} };
+
+// Discard words that have letters which are not in the grid.
+//
+set<string> reduceDictionary(set<string> dictionary, set<char> gridLetters) {
+	set<string> result;
+
+	for (set<string>::iterator it = dictionary.begin(); it != dictionary.end(); ++it) {
+		int cnt = 0;
+		string word = *it;
+		transform(word.begin(), word.end(), word.begin(), ::tolower);
+
+		for (int i = 0; i < word.length(); i++) {
+			if (gridLetters.find(word[i]) == gridLetters.end())
+				break;
+			cnt++;
+		}
+
+		if (word.length() == cnt)
+			result.insert(word);
+	}
+
+	return result;
+}
+
+// This allows the search function to abort search paths which can never result in words.
+//
+set<string> generateWordsPrefixes(set<string> dictionary) {
+	set<string> result;
+
+	for (set<string>::iterator it = dictionary.begin(); it != dictionary.end(); ++it) {
+		string word = *it;
+		for (int i = 0; i <= word.length(); i++) {
+			result.insert(word.substr(0, i));
+		}
+	}
+
+	return result;
+}
+
+set<char> makeGridSet(GridCell grid[][3]) {
+	set<char> result;
+	for (int i = 0; i < 3; i++)
+		for (int j = 0; j < 3; j++) {
+			result.insert(grid[i][j].letter);
+		}
+	return result;
+}
+
+bool isPrefix(set<string> prefixes, string word) {
+	return prefixes.find(word) != prefixes.end();
+}
+
+bool isWord(set<string> dictionary, string word) {
+	return dictionary.find(word) != dictionary.end();
+}
+
+bool check(int i, int j) {
+	return i >= 0 && i < 3 && j >= 0 && j < 3 && !grid[i][j].visited;
+}
+
+void findWords(set<string> dictionary, set<string> prefixes, vector<string>& matches, int i, int j, vector<char> generatedWord) {
+	bool validWord = true;
+
+	generatedWord.push_back(grid[i][j].letter);
+	
+	string word(generatedWord.begin(), generatedWord.end());
+
+	if (!isPrefix(prefixes, word)) {
+		return;
+	}
+
+	if (isWord(dictionary, word)) {
+		matches.push_back(word);
+	}
+
+	grid[i][j].visited = true;
+	vector<char> newWord(generatedWord);
+
+	for (int k = 0; k < 8; k++) {
+		bool validToMove = check(i + difference[k].dx, j + difference[k].dy);
+		if(validToMove)
+			findWords(dictionary, prefixes, matches, i + difference[k].dx, j + difference[k].dy, newWord);
+	}
+
+	grid[i][j].visited = false;
+	generatedWord.pop_back();
+}
+
+// If word can be formed in the grid, it will be included in resulting vector.
+//
+vector<string> findAllWordsFormDictionaryInTheGrid(set<string> dictionary, GridCell grid[][3]) {
```

Comment:
Your code gets `dictionary` as a set of strings, meaning that anyone who uses your code will have to implement their dictionary as a set of strings. What if you want to make it easy for the user to change the implementation of dictionary without having to modify this code?
Exactly for that, Object Oriented Programming comes in!
You can create a dictionary class, and *encapsulate* the implementation in the class, providing a simple interface:
```C++
class Dictionary {
  private:
    // Some fields, for example:
    std::set<string> words;
  public:
    // constructor
    Dictionary(const std::set<string>& words) ...

    // Interface functions, the any user of this function will use these functions only!
    bool isPrefix(const string& str) { ... }
    bool isWord(const string& str) { ... }
};
```

Then in your code, you can get the dictionary as a parameter:
vector<string> findAllWords(const Dictionary& dictionary, ...) {
   ...
   if (dictionary.isPrefix(word)) ...  // Example usage.
}

This case, your solution only uses `isPrefix` and `isWord` functions, and does not rely on the implementation of Dictionary. Now the next programmer which comes to change or use this code (which can be you in the future or someone else) can change the dictionary implementation (for exmple, if someone wants to improve its complexity), without the need to change anything of the "find words in grid" algorithm. They only have to make sure that at the end, `isPrefix` and `isWord` behave the same.
Easy! Portable! Encapsulated! Object Oriented! :-0